File: dart/lib/src/eip712.dart
/// Production EIP-712 encoding for AIMarketEscrow debit authorizations.
library;

import 'dart:typed_data';

import 'package:web3dart/crypto.dart';

const String eip712DomainType =
    'EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)';

const String debitTypehashHeader =
    'DebitAuthorization(bytes32 channelId,address hub,address token,uint256 amount,bytes32 receiptId,uint256 nonce,uint256 deadline)';

const String escrowContractName = 'AIMarketEscrow';
const String escrowContractVersion = '1';

Uint8List _keccak256(Uint8List data) => keccak256(data);

Uint8List _keccak256Utf8(String data) => _keccak256(Uint8List.fromList(data.codeUnits));

Uint8List _parseHex32(String value) {
  final hex = value.startsWith('0x') ? value.substring(2) : value;
  final bytes = List<int>.generate(hex.length ~/ 2, (i) {
    return int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
  });
  if (bytes.length != 32) {
    throw ArgumentError('expected 32-byte hex, got ${bytes.length}');
  }
  return Uint8List.fromList(bytes);
}

Uint8List _encodeAddress(String address) {
  final hex = address.startsWith('0x') ? address.substring(2) : address;
  final bytes = List<int>.generate(hex.length ~/ 2, (i) {
    return int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
  });
  final out = Uint8List(32);
  out.setRange(32 - bytes.length, 32, bytes);
  return out;
}

Uint8List _encodeU256(BigInt value) {
  final out = Uint8List(32);
  final bytes = _bigIntToBytes(value);
  out.setRange(32 - bytes.length, 32, bytes);
  return out;
}

List<int> _bigIntToBytes(BigInt value) {
  if (value == BigInt.zero) return [0];
  final bytes = <int>[];
  var v = value;
  while (v > BigInt.zero) {
    bytes.insert(0, (v & BigInt.from(0xff)).toInt());
    v >>= 8;
  }
  return bytes;
}

Uint8List _abiEncode(List<Uint8List> words) {
  final out = BytesBuilder();
  for (final word in words) {
    out.add(word);
  }
  return out.toBytes();
}

/// Compute EIP-712 domain separator for AIMarketEscrow.
Uint8List buildDomainSeparator({
  required int chainId,
  required String verifyingContract,
}) {
  final domainTypeHash = _keccak256Utf8(eip712DomainType);
  final nameHash = _keccak256Utf8(escrowContractName);
  final versionHash = _keccak256Utf8(escrowContractVersion);
  return _keccak256(_abiEncode([
    domainTypeHash,
    nameHash,
    versionHash,
    _encodeU256(BigInt.from(chainId)),
    _encodeAddress(verifyingContract),
  ]));
}

/// Compute EIP-712 digest for a debit authorization.
Uint8List computeDebitDigest({
  required String channelId,
  required String hub,
  required String token,
  required BigInt amount,
  required String receiptId,
  required BigInt nonce,
  required int deadline,
  int chainId = 8453,
  String verifyingContract = '0x0000000000000000000000000000000000000000',
}) {
  final typeHash = _keccak256Utf8(debitTypehashHeader);
  final structHash = _keccak256(_abiEncode([
    typeHash,
    _parseHex32(channelId),
    _encodeAddress(hub),
    _encodeAddress(token),
    _encodeU256(amount),
    _parseHex32(receiptId),
    _encodeU256(nonce),
    _encodeU256(BigInt.from(deadline)),
  ]));
  final domain = buildDomainSeparator(
    chainId: chainId,
    verifyingContract: verifyingContract,
  );
  final buf = BytesBuilder();
  buf.addByte(0x19);
  buf.addByte(0x01);
  buf.add(domain);
  buf.add(structHash);
  return _keccak256(buf.toBytes());
}

File: rust/src/eip712.rs
//! Production EIP-712 encoding for AIMarketEscrow debit authorizations.

use tiny_keccak::{Hasher, Keccak};

pub const EIP712_DOMAIN_TYPE: &str =
    "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)";

pub const DEBIT_TYPEHASH_HEADER: &str =
    "DebitAuthorization(bytes32 channelId,address hub,address token,uint256 amount,bytes32 receiptId,uint256 nonce,uint256 deadline)";

pub const ESCROW_CONTRACT_NAME: &str = "AIMarketEscrow";
pub const ESCROW_CONTRACT_VERSION: &str = "1";

#[derive(Debug, Clone)]
pub struct DebitDigestParams<'a> {
    pub channel_id: &'a str,
    pub hub: &'a str,
    pub token: &'a str,
    pub amount: u128,
    pub receipt_id: &'a str,
    pub nonce: u128,
    pub deadline: u128,
    pub chain_id: u64,
    pub verifying_contract: &'a str,
}

pub fn keccak256(data: &[u8]) -> [u8; 32] {
    let mut out = [0u8; 32];
    let mut h = Keccak::v256();
    h.update(data);
    h.finalize(&mut out);
    out
}

fn parse_hex32(value: &str) -> [u8; 32] {
    let hex = value.strip_prefix("0x").unwrap_or(value);
    let bytes = hex::decode(hex).expect("invalid 32-byte hex");
    assert_eq!(bytes.len(), 32, "expected 32 bytes, got {}", bytes.len());
    let mut out = [0u8; 32];
    out.copy_from_slice(&bytes);
    out
}

fn encode_address(value: &str) -> [u8; 32] {
    let hex = value.strip_prefix("0x").unwrap_or(value);
    let bytes = hex::decode(hex).expect("invalid address hex");
    assert!(bytes.len() <= 20, "address longer than 20 bytes");
    let mut out = [0u8; 32];
    out[32 - bytes.len()..].copy_from_slice(&bytes);
    out
}

fn encode_u256(value: u128) -> [u8; 32] {
    let mut out = [0u8; 32];
    out[16..].copy_from_slice(&value.to_be_bytes());
    out
}

fn abi_encode(words: &[[u8; 32]]) -> Vec<u8> {
    words.iter().flat_map(|w| w.iter().copied()).collect()
}

pub fn domain_separator(chain_id: u64, verifying_contract: &str) -> [u8; 32] {
    let domain_type_hash = keccak256(EIP712_DOMAIN_TYPE.as_bytes());
    let name_hash = keccak256(ESCROW_CONTRACT_NAME.as_bytes());
    let version_hash = keccak256(ESCROW_CONTRACT_VERSION.as_bytes());
    keccak256(&abi_encode(&[
        domain_type_hash,
        name_hash,
        version_hash,
        encode_u256(chain_id as u128),
        encode_address(verifying_contract),
    ]))
}

pub fn debit_struct_hash(params: &DebitDigestParams<'_>) -> [u8; 32] {
    let type_hash = keccak256(DEBIT_TYPEHASH_HEADER.as_bytes());
    keccak256(&abi_encode(&[
        type_hash,
        parse_hex32(params.channel_id),
        encode_address(params.hub),
        encode_address(params.token),
        encode_u256(params.amount),
        parse_hex32(params.receipt_id),
        encode_u256(params.nonce),
        encode_u256(params.deadline),
    ]))
}

/// EIP-712 digest: `keccak256(0x1901 || domainSeparator || structHash)`.
pub fn compute_debit_digest(params: &DebitDigestParams<'_>) -> [u8; 32] {
    let domain = domain_separator(params.chain_id, params.verifying_contract);
    let struct_hash = debit_struct_hash(params);
    let mut buf = Vec::with_capacity(2 + 64);
    buf.extend_from_slice(&[0x19, 0x01]);
    buf.extend_from_slice(&domain);
    buf.extend_from_slice(&struct_hash);
    keccak256(&buf)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn digest_is_deterministic() {
        let params = DebitDigestParams {
            channel_id: "0x0000000000000000000000000000000000000000000000000000000000000001",
            hub: "0x000000000000000000000000000000000000bEEF",
            token: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
            amount: 5_000_000,
            receipt_id: "0x0000000000000000000000000000000000000000000000000000000000001234",
            nonce: 0,
            deadline: 2_000_000_000,
            chain_id: 31_337,
            verifying_contract: "0x5FbDB2315678afecb367f032d93F642f64180aa3",
        };
        let a = compute_debit_digest(&params);
        let b = compute_debit_digest(&params);
        assert_eq!(a, b);
    }
}
